Merge Branch-DG-DOB Hong Liang with Master (#110)
AY2223S1-CS2103-F09-2/tp: docs/DeveloperGuide.md

contact list. Error message is displayed on the GUI subsequently.
</div>

### Add Date Of Birth

> The Add Date Of Birth feature allows users to add a date of birth field (format: dd/mm/yyyy) to a person in the contact list. It is performed as a part of `addPersonCommand#execute()`.

Reviewer: contact list is a term we did not define in our DG and its we do have a class or field for it. Perhaps, we could use the term 'application' or 'addressbook' in place of the contact list and it'll be less confusing for first time readers of our DG/ users of our application?  

Also, if you're changing it from contact list, there are a few more instances of the phrase below which you will have to change too!

These operations are exposed in the `Model` interface as the method `Model#addPerson()`, which calls
`AddressBook#addPerson()` which calls `UniquePersonList#add()` to add a new person in the person list
stored in AddressBook.

The following sequence diagram shows the methods calls related to add person operation:

![AddPersonSequenceDiagram](images/AddPersonSequenceDiagram.png)

<div markdown="span" class="alert alert-info">:information_source: **Note:** `cmd` in the diagram represents the add
person command text entered by user. The specific `UniquePersonList` operations are not shown in the diagram
for simplicity.
</div>

> The following activity diagram shows what happens when a user executes a new add command:

Reviewer: To standardize with line 284, we should call it - new add person command (we have yet to rename it but it'll be done this iteration so using the term 'add person' is preferable to 'add' in this DG)

![AddPersonActivityDiagram](images/AddPersonActivityDiagram_DOB.png)

<div markdown="span" class="alert alert-info">:information_source: **Note:** Only the
activities related to date of birth field are considered and shown in this activity diagram.
</div>

<div markdown="span" class="alert alert-info">:information_source: **Note:**Parser exceptions are thrown and caught if

> date of birth field is not provided in the command, or the date of birth is not of valid format; Duplicated person exception is

Reviewer: The exception class is called DuplicatePersonException so duplicate may be better represent it than duplicated. 
```suggestion
date of birth field is not provided in the command, or the date of birth is not of valid format; Duplicate person exception is
```

thrown if the person to add already exists in the contact list. Error message is displayed on the GUI subsequently.
</div>

#### Design considerations:

**Aspect: Whether date of birth field should be optional for a person:**

* **Alternative 1 (current choice):** Compulsory date of birth field:

>     * Pros: It is a more logical implementation because date of birth is a common attribute for all persons,
>   similar to name, address, etc, which are also compulsory field for persons in contact list.

Reviewer: Minor edit: should be plural
```suggestion
    * Pros: It is a more logical implementation because date of birth is a common attribute for all persons,
  similar to name, address, etc, which are also compulsory fields for persons in contact list.
```

    * Cons: It is less flexible in cases where the user is missing the date of birth field for a contact,
    but wants to add the contact anyways, as the user is not able to leave the date of birth field blank.

* **Alternative 2:** Optional date of birth field:
    * Pros: It is a more flexible implementation, since the user has the choice to set a date of birth,
    or leave it empty.

>     * Cons: It is less logical since date of birth is a common attribute for a person.

Reviewer: Might be better phrased this way
```suggestion
    * Cons: It is less logical since date of birth is a common attribute for all persons in the contact list.
```

It is performed as a part of `editPersonCommand#execute()`.

These operations are exposed in the `Model` interface as the method `Model#setPerson()`, which calls
`AddressBook#setPerson()` which calls `UniquePersonList#setPerson()` to replace an existing person with a new person
object with edited fields in the person list stored in AddressBook.

The following sequence diagram shows the methods calls related to edit person operation:

![EditPersonSequenceDiagram](images/EditPersonSequenceDiagram.png)

<div markdown="span" class="alert alert-info">:information_source: **Note:** `cmd` in the diagram represents the edit
person command text entered by user; the `setPerson(P1, P2)` method replaces person P1 with person P2 in the person list
in the model. The specific `UniquePersonList` operations are not shown in the diagram for simplicity.
</div>

The following activity diagram shows what happens when a user executes a new edit command:

![EditPersonActivityDiagram](images/EditPersonActivityDiagram_DOB.png)

<div markdown="span" class="alert alert-info">:information_source: **Note:** Only the

> activities related to date of birth field are considered and shown in this activity diagram. All fields are considered optional

Reviewer: "All fields are considered optional" --> Not entirely true as at least 1 field must be provided and the others may then be optional.
